### PR TITLE
0.4.2 bump crypto 0.22

### PR DESCRIPTION
**Pull request by semenov-vladyslav**
# Description of change

Update 0.4.2 branch:

- use snapshot v3 age format;
- added Zeroizing;
- update to crypto 0.21 (dev);
- impl Secp256k1 ECDSA procedures: PublicKey, EvmAddress, Sign;
- support curves with SLIP10;
- updated BIP39 impl.

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes issue #`.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: client/src/actors/client.rs [hunk 1]
```diff
@@ -27,9 +27,63 @@ use core::{
...
+    pub fn deserialize<'de, D>(d: D) -> Result<bip39::Mnemonic, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::Deserialize;
+        String::deserialize(d).map(bip39::Mnemonic::from)
+    }
+}
+
+pub mod serde_bip39_passphrase {
```
Comment (thibault-martinez) [inline client/src/actors/client.rs]: Can you use only one like in the other PR?

File: client/src/lib.rs [hunk 2]
```diff
@@ -59,6 +59,8 @@ pub use engine::snapshot::{
 
 pub use engine::vault::RecordHint;
```
Comment (thibault-martinez) [inline client/src/lib.rs]:
```suggestion
pub use engine::{self, vault::RecordHint};
```
Just in case you didn't know the syntax :)

Comment (semenov-vladyslav) [reply client/src/lib.rs]: Right, fixed.

File: client/src/actors/client.rs (hunk 1 above)
Comment (semenov-vladyslav) [reply client/src/actors/client.rs]: Forgot to update it. Fixed.

File: engine/src/snapshot/migration/v3.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+/// Key size for the ephemeral key
+const KEY_SIZE: usize = 32;
+/// Key type alias.
+pub type Key = [u8; KEY_SIZE];
+
+const VERSION_V3: [u8; 2] = [0x3, 0x0];
+
+#[allow(dead_code)]
+pub(crate) fn read<I: Read>(input: &mut I, password: &Key, _associated_data: &[u8]) -> Result<Vec<u8>, Error> {
```
Comment (thibault-martinez) [inline engine/src/snapshot/migration/v3.rs]: ` _associated_data` has been removed in 2.0, do we want to keep it here ?

Comment (semenov-vladyslav) [reply engine/src/snapshot/migration/v3.rs]: Sure. Removed.

File: engine/src/snapshot/migration/v3.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    output.write_all(&age[..])?;
+    Ok(())
+}
+
+/// Atomically encrypt, add magic and version bytes as file-header, and [`write`][self::write] the specified
+/// plaintext to the specified path.
+///
+/// This is achieved by creating a temporary file in the same directory as the specified path (same
+/// filename with a salted suffix). This is currently known to be problematic if the path is a
+/// symlink and/or if the target path resides in a directory without user write permission.
```
Comment (thibault-martinez) [inline engine/src/snapshot/migration/v3.rs]: Wrong comment

Comment (semenov-vladyslav) [reply engine/src/snapshot/migration/v3.rs]: Removed.

File: client/src/actors/internal.rs [hunk 5]
```diff
@@ -676,11 +700,12 @@ impl Receive<InternalMsg> for InternalActor<Provider> {
...
                             }
                             raw.truncate(32);
-                            let mut bs = [0; 32];
+                            let mut bs = Zeroizing::new([0; 32]);
                             bs.copy_from_slice(&raw);
-                            let sk = ed25519::SecretKey::from_le_bytes(bs).expect(line_error!());
+                            let sk = ed25519::SecretKey::from_bytes(&bs);
 
                             let sig = sk.sign(&msg);
+                            // sk.zeroize();
```
Comment (thibault-martinez) [inline client/src/actors/internal.rs]: ?

Comment (semenov-vladyslav) [reply client/src/actors/internal.rs]: Removed. sk is ZeroizeOnDrop.

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**